nameresolution/consul/watcher.go

```diff
@@ -0,0 +1,342 @@
+	changedServices := make(map[string]struct{})
+
+	// get changed services
+	for newKey, newValue := range newResult {
+		if oldValue, ok := p.lastResult[newKey]; ok && newValue == oldValue {
+			continue
+		}
+
+		changedServices[newKey.serviceName] = struct{}{}
+	}
+
+	for oldKey := range p.lastResult {
+		if _, ok := newResult[oldKey]; !ok {
+			changedServices[oldKey.serviceName] = struct{}{}
+		}
+	}
+
+	return changedServices
+}
+
+func (p *watchPlan) buildServiceNameFilter(services []string) {
+	var nameFilters = make([]string, len(services))
+
+	for i, v := range services {
+		nameFilters[i] = fmt.Sprintf("ServiceName==\"%s\"", v)
+	}
+
+	filter := fmt.Sprintf("(%s)", strings.Join(nameFilters, " or "))
+
+	if len(p.configuredQueryFilter) < 1 {
+		p.options.Filter = filter
+	} else {
+		p.options.Filter = fmt.Sprintf("%s and %s", p.configuredQueryFilter, filter)
+	}
+}
+
+func (r *resolver) watch(p *watchPlan, services []string, ctx context.Context) (blockingParamVal, consul.HealthChecks, error, bool) {
+	p.options = p.options.WithContext(ctx)
+
+	if p.lastParamVal != nil {
+		p.options.WaitIndex = uint64(p.lastParamVal.(waitIndexVal))
+	}
+
+	p.buildServiceNameFilter(services)
+
+	checks, meta, err := r.client.Health().State(consul.HealthAny, p.options)
+
+	if err != nil {
+		// if it failed during long poll try again with no wait
+		if p.options.WaitIndex != uint64(0) {
+			p.options.WaitIndex = 0
+			checks, meta, err = r.client.Health().State(consul.HealthAny, p.options)
+		}
+
+		if err != nil {
+			// if the context was canceled
+			if errors.Is(err, context.Canceled) {
+				return nil, nil, nil, true
+			}
+
+			// if it failed with no wait and plan is not expired
+			if p.options.WaitIndex == uint64(0) && !p.expired {
+				p.lastResult = nil
+				p.expired = true
+				r.registry.expireAll()
+			}
+
+			return nil, nil, err, false
+		}
+	}
+
+	p.expired = false
+
+	return waitIndexVal(meta.LastIndex), checks, err, false
+}
+
+func (r *resolver) runWatchPlan(p *watchPlan, services []string, ctx context.Context, watchTask chan bool) {
+	defer func() {
+		recover()
+		// complete watch task
+		watchTask <- true
+	}()
+
+	// invoke blocking call
+	blockParam, result, err, canceled := r.watch(p, services, ctx)
+
+	// if the ctx was canceled then do nothing
+	if canceled {
+		return
+	}
+
+	// handle an error in the watch function
+	if err != nil {
+		// always 0 on err so query is forced to return
+		p.lastParamVal = waitIndexVal(0)
+
+		// perform an exponential backoff
+		p.failures++
+
+		retry := retryInterval * time.Duration(p.failures*p.failures)
+		if retry > maxBackoffTime {
+			retry = maxBackoffTime
+		}
+
+		// pause watcher routine until ctx is canceled or retry timer finishes
+		r.logger.Errorf("consul service-watcher error: %v, retry in %v", err, retry)
+		sleepTimer := time.NewTimer(retry)
+		select {
+		case <-ctx.Done():
+			sleepTimer.Stop()
+			r.logger.Debugf("consul service-watcher retry throttling canceled")
```

Comment:
If there's no formatting, don't use `f`.

```suggestion
			r.logger.Debug("consul service-watcher retry throttling canceled")
```